src/load_gdm.cpp

```diff
@@ -0,0 +1,461 @@
+
+	if ((ordersPos < sizeof(FILEHEADERGDM)) || (ordersPos >= dwMemLength) || (ordersPos + nOrders > dwMemLength) ||
+	    (patternsPos < sizeof(FILEHEADERGDM)) || (patternsPos >= dwMemLength) ||
+	    (samplesPos < sizeof(FILEHEADERGDM)) || (samplesPos >= dwMemLength) ||
+	    (samplesPos + sizeof(SAMPLEGDM) * pfh->nSamples > dwMemLength) ||
+	    (sampleDataPos < sizeof(FILEHEADERGDM)) || (sampleDataPos >= dwMemLength))
+		return FALSE;
+
+	// Most GDMs were converted from S3M and BWSB generally behaves like an S3M
+	// player, so assuming S3M behavior and quirks is a fairly safe bet.
+	m_nType = MOD_TYPE_GDM | MOD_TYPE_S3M;
+	m_nMinPeriod = 64;
+	m_nMaxPeriod = 32767;
+	m_nDefaultGlobalVolume = pfh->globalvol << 2;
+	m_nDefaultTempo = pfh->default_bpm;
+	m_nDefaultSpeed = pfh->default_speed;
+	m_nChannels = 0;
+	m_nSamples = nSamples;
+
+	// Get initial panning.
+	for (UINT i = 0; i < 32; i++)
+	{
+		if (pfh->panning[i] < 16)
+		{
+			ChnSettings[i].nPan = (pfh->panning[i] << 4) + 8;
+		}
+		// TODO 16=surround
+	}
+
+	// Title and message.
+	memcpy(m_szNames[0], pfh->name, 32);
+	m_szNames[0][31] = '\0';
+
+	if ((messagePos < dwMemLength) && (messagePos + messageLen < dwMemLength))
+	{
+		m_lpszSongComments = new char[messageLen + 1];
+		memcpy(m_lpszSongComments, lpStream + messagePos, messageLen);
+		m_lpszSongComments[messageLen] = '\0';
+	}
+
+	// Samples.
+	psmp = (const SAMPLEGDM *)(lpStream + samplesPos);
+	for (UINT nins = 0; nins < nSamples; nins++)
+	{
+		const SAMPLEGDM &smp = psmp[nins];
+		MODINSTRUMENT &ins = Ins[nins + 1];
+
+		DWORD len = fixu32(smp.length);
+		DWORD loopstart = fixu32(smp.loopstart);
+		DWORD loopend = fixu32(smp.loopend);
+
+		UINT flags = smp.flags;
+
+		// Note: BWSB and 2GDM don't support LZW, stereo samples, sample panning.
+		if (flags & SAMPLEGDM::S_16BIT)
+		{
+			sflags[nins] = RS_PCM16U;
+			// Due to a 2GDM bug, the sample size is halved.
+			// (Note BWSB doesn't even check for these anyway.)
+			len /= 2;
+		}
+		else
+			sflags[nins] = RS_PCM8U;
+
+		if (len > MAX_SAMPLE_LENGTH) len = MAX_SAMPLE_LENGTH;
+		if (loopend > len) loopend = len;
+		if (loopstart > loopend) loopstart = loopend = 0;
+
+		ins.nLength = len;
+		ins.nLoopStart = loopstart;
+		ins.nLoopEnd = loopend;
+		if (loopend && (flags & SAMPLEGDM::S_LOOP)) ins.uFlags |= CHN_LOOP;
+
+		ins.nC4Speed = fixu16(smp.c4rate);
+		ins.nVolume = (flags & SAMPLEGDM::S_VOLUME) && (smp.volume <= 64) ? (smp.volume << 2) : 256;
+		ins.nGlobalVol = 64;
+		ins.nPan = 128;
+
+		memcpy(m_szNames[nins], smp.name, 32);
+		m_szNames[nins][31] = '\0';
+
+		memcpy(ins.name, smp.dosname, 12);
+		ins.name[13] = '\0';
+	}
+
+	// Order table.
+	memcpy(Order, lpStream + ordersPos, pfh->nOrders + 1);
+
+	// Scan patterns to get pattern row counts and the real module channel count.
+	// Also do bounds checks to make sure the patterns can be safely loaded.
+	pos = patternsPos;
+	for (UINT npat = 0; npat < nPatterns && pos + 2 <= dwMemLength; npat++)
```

Comment:
I did the following to fix build using really old MSVC (Yeah, I know it's blah..
there are 2 or 3 other places in mainstream tree which need similar touches.)
```diff
--- load_gdm.cpp~
+++ load_gdm.cpp
@@ -232,2 +232,3 @@ BOOL CSoundFile::ReadGDM(const BYTE *lpS
 	DWORD pos;
+	UINT npat;
 
@@ -340,3 +341,3 @@ BOOL CSoundFile::ReadGDM(const BYTE *lpS
 	pos = patternsPos;
-	for (UINT npat = 0; npat < nPatterns && pos + 2 <= dwMemLength; npat++)
+	for (npat = 0; npat < nPatterns && pos + 2 <= dwMemLength; npat++)
 	{
@@ -389,3 +390,3 @@ BOOL CSoundFile::ReadGDM(const BYTE *lpS
 	pos = patternsPos;
-	for (UINT npat = 0; npat < nPatterns && pos + 2 <= dwMemLength; npat++)
+	for (npat = 0; npat < nPatterns && pos + 2 <= dwMemLength; npat++)
 	{
```

Reply:
I've applied this patch, thanks.